milli/src/update/new/mod.rs

```diff
@@ -0,0 +1,619 @@
+                            document: all_obkv_to_json(&document, fields_ids_map)?,
+                        })
+                    }
+                }?;
+
+                let insertion = Insertion::create(docid, external_docid, document);
+                Ok(Some(DocumentChange::Insertion(insertion)))
+            })
+        }
+    }
+
+    pub struct UpdateByFunctionIndexer;
+
+    /// TODO return stats
+    /// TODO take the rayon ThreadPool
+    pub fn index<PI>(
+        wtxn: &mut RwTxn,
+        index: &Index,
+        pool: &ThreadPool,
+        document_changes: PI,
+    ) -> Result<()>
+    where
+        PI: IntoParallelIterator<Item = Result<DocumentChange>> + Send,
+        PI::Iter: Clone,
+    {
+        let (merger_sender, writer_receiver) = merger_writer_channels(100);
+        let ExtractorsMergerChannels { merger_receiver, deladd_cbo_roaring_bitmap_sender } =
+            extractors_merger_channels(100);
+
+        thread::scope(|s| {
+            thread::Builder::new().name(S("indexer-extractors")).spawn_scoped(s, || {
+                pool.in_place_scope(|_s| {
+                    document_changes.into_par_iter().for_each(|_dc| ());
+                })
+            })?;
+
+            // TODO manage the errors correctly
+            thread::Builder::new().name(S("indexer-merger")).spawn_scoped(s, || {
+                let rtxn = index.read_txn().unwrap();
+                merge_grenad_entries(merger_receiver, merger_sender, &rtxn, index).unwrap()
+            })?;
+
+            // TODO Split this code into another function
+            for operation in writer_receiver {
+                let database = operation.database(index);
+                match operation {
+                    WriterOperation::WordDocids(operation) => match operation {
+                        EntryOperation::Delete(e) => database.delete(wtxn, e.entry()).map(drop)?,
+                        EntryOperation::Write(e) => database.put(wtxn, e.key(), e.value())?,
+                    },
+                    WriterOperation::Document(e) => database.put(wtxn, &e.key(), e.content())?,
+                }
+            }
+
+            Ok(())
+        })
+    }
+
+    enum Operation {
+        Write(RoaringBitmap),
+        Delete,
+        Ignore,
+    }
+
+    /// A function that merges the DelAdd CboRoaringBitmaps with the current bitmap.
+    fn merge_cbo_bitmaps(
+        current: Option<&[u8]>,
+        del: Option<&[u8]>,
+        add: Option<&[u8]>,
+    ) -> Result<Operation> {
+        let bitmap = match current {
+            Some(current_bitmap_bytes) => {
+                let bitmap_without_del = match del {
+                    Some(del_bytes) => {
+                        let del_bitmap = CboRoaringBitmapCodec::deserialize_from(del_bytes)?;
+                        CboRoaringBitmapCodec::intersection_with_serialized(
+                            current_bitmap_bytes,
+                            &del_bitmap,
+                        )?
```

Comment:
This is wrong. We must do a difference not an intersection. We will have to materialize all three bitmaps at least for now.

Reply:
Fixed in bb885a5